src/components/character-count/character-count.js

```diff
@@ -0,0 +1,226 @@
+import '../../vendor/polyfills/Function/prototype/bind'
+import '../../vendor/polyfills/Event' // addEventListener and event.target normaliziation
+
+function CharacterCount ($module) {
+  this.$module = $module
+  this.$textarea = $module.querySelector('.js-character-count')
+}
+
+CharacterCount.prototype.defaults = {
+  characterCountAttribute: 'data-maxlength',
+  wordCountAttribute: 'data-maxwords'
+}
+
+// Initialize component
+CharacterCount.prototype.init = function (options) {
+  // Check for module
+  var $module = this.$module
+  var $textarea = this.$textarea
+  if (!$textarea) {
+    return
+  }
+
+  // Read options set using dataset ('data-' values) and merge them with options
+  var dataset = this.getDataset($module)
+  this.options = this.merge(options || {}, dataset)
+
+  // Determine the limit attribute (characters or words)
+  var countAttribute = (this.options.maxwords) ? this.defaults.wordCountAttribute : this.defaults.characterCountAttribute
+
+  // Set the element limit
+  if ($module.getAttribute) {
+    this.maxLength = $module.getAttribute(countAttribute)
+  }
+
+  // Generate and reference message
+  var boundCreateCountMessage = this.createCountMessage.bind(this)
+  this.countMessage = boundCreateCountMessage()
+
+  // If there's a maximum length defined and the count message was successfully applied
+  if (this.maxLength && this.countMessage) {
+    // Replace maxlength attribute with data-maxlength to avoid hard limit
+    $module.setAttribute('maxlength', '')
+    $module.setAttribute('data-maxlength', $module.maxLength)
```

Comment:
I'm skeptic about moving these to the textarea instead because in my opinion they are being used at the component level not at the textarea element level. Also we should keep it flexible and easier to extend in the future so that the field could be an input or a textarea.

Reply:
My bad I assumed `maxlength` was set on `$textarea`. The maxlength value should be from the character count instance, not from the module as `$module.maxLength` is `undefined`. Think it should be:

```
$module.setAttribute('data-maxlength', this.maxLength)
```